packages/core/src/RepoSync.ts

```diff
@@ -202,255 +163,96 @@ export class RepoSync {
-    await this.maybeRequestChanges(documentId, clock)
-
-    // update our clock
-    this.updateClock(documentId, ours, clock)
   }
 
-  /**
-   * Sends a hello message including our document count
-   */
+  /** Sends a hello message including our document count */
   private async sendHello() {
     const documentCount = this.repo.count
     this.log('sending hello', documentCount)
     this.send({ type: message.HELLO, documentCount })
   }
 
-  /**
-   * Checks whether peer has more recent information than we do; if so, requests changes
-   * @param documentId
-   * @param theirClock
-   */
-  private async maybeRequestChanges(
-    documentId: string,
-    theirClock: Clock | Promise<Clock> = this.getClockFromDoc(documentId)
-  ) {
-    this.log('maybeRequestChanges', documentId)
-    const ourClock = this.getOurClock(documentId)
-    theirClock = await theirClock
-    if (isMoreRecent(theirClock, ourClock)) this.advertise(documentId, theirClock)
-  }
-
-  /**
-   * Checks whether we have more recent information than they do; if so, sends changes
-   * @param documentId
-   */
+  /** Checks whether we have more recent information than they do; if so, sends changes */
   private async maybeSendChanges(documentId: string) {
     const theirClock = this.getTheirClock(documentId)
-    if (theirClock === undefined) return
-    const ourDoc = await this.repo.get(documentId)
-    const changes = _A.getMissingChanges(ourDoc, (theirClock as unknown) as A.Clock)
-    if (changes.length > 0) await this.sendChanges(documentId, changes)
+    const ourClock = this.getOurClock(documentId)
+    if (isMoreRecent(ourClock, theirClock)) {
+      const ourDoc = await this.repo.get(documentId)
+      if (ourDoc === undefined) return
+      const changes = theirClock
+        ? getMissingChanges(ourDoc, theirClock)
+        : A.getChanges(A.init(), ourDoc)
+      if (changes.length > 0) await this.sendChanges(documentId, changes)
+    }
   }
 
-  /**
-   * Sends a changeset to our peer, bringing them up to date with our latest info
-   * @param documentId
-   * @param changes
-   */
+  /** Sends a changeset to our peer, bringing them up to date with our latest info */
   private async sendChanges(documentId: string, changes: A.Change[]) {
     this.log('sendChanges', documentId)
-    const clock = await this.getClockFromDoc(documentId)
-    this.send({
-      type: message.SEND_CHANGES,
-      documentId,
-      clock: clock.toJS() as Clock,
-      changes,
-    })
-    this.updateClock(documentId, ours)
-  }
-
-  /**
-   * Informs our peer that we have a specific version of a document, so they can see if they have an
-   * older version (in which case they will request changes) or a newer version (in which case they
-   * will send changes)
-   * @param documentId
-   * @param [_clock]
-   */
-  private async advertise(documentId: string, _clock: Clock = this.getOurClock(documentId)) {
-    this.log('advertise', documentId)
```

Comment:
since we advertise all on first connection, there's never really a need to advertise a single doc. If it's a new doc or it changes, we know the changes they need and we just send them.